src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs

```diff
@@ -508,14 +525,30 @@ public Task RelayTransaction(uint256 hash)
             // find all behaviours then start an exclusive task
             // to add the hash to each local collection
             IEnumerable<MempoolBehavior> behaviours = peers.Select(s => s.Behavior<MempoolBehavior>());
-            this.logger.LogTrace("(-)");
-            return this.manager.MempoolLock.WriteAsync(() =>
+            return this.manager.MempoolLock.WriteAsync(() => 
             {
                 foreach (MempoolBehavior mempoolBehavior in behaviours)
                 {
+                    this.logger.LogTrace("Attempting to relaying transaction ID '{0}' to peer '{1}'.", hash, mempoolBehavior?.AttachedPeer.RemoteSocketEndpoint);
                     if (mempoolBehavior?.AttachedPeer.PeerVersion.Relay ?? false)
-                        if (!mempoolBehavior.filterInventoryKnown.ContainsKey(hash))
-                            mempoolBehavior.inventoryTxToSend.TryAdd(hash, hash);
+                    {
+                        using (mempoolBehavior.asyncLock.Lock())
```

Comment:
Please don't use private members of other instances of the same class. 
Expose a public interface for handling such situations.

Reply:
Why not? this is a very general suggestion you make @noescape00 stick to the review.